# <p style='text-align: center;'> Descriptive Statistics using Numpy </p>

- On the object of ndarray, we are going to perform Statistical Operations.


- we can perform on the entire matrix and we can also perform on column-wise (axis=0) & Row-wise (axis=1).


## sum() : -
- This function is used to compute the sum of all elements, the sum of each row, and the sum of each column of a given array.


- There is the following syntax of numpy.sum() function :


- <b> numpy.sum(arr, axis=None, dtype=None)          
            
            
- Where, 
- i) arr : This is the source array whose elements we want to sum.


- ii) axis (optional) : This parameter defines the axis along which a sum is performed. The default axis is None,column-wise          (axis=0) & Row-wise (axis=1).


- iii) dtype (optional) : This parameter defines the type of the accumulator and the returned array in which the elements are 
summed. By default, the dtype of arr is used unless arr has an integer dtype of less precision than                            the default platform integer.  

In [2]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([154, 139, 154, 192, 180, 140, 154, 155, 192])

#  Calculate the sum
# Print the result
np.sum(x)    # It will print the sum of all the elements of the array.

1460

<b> In the above example the sum of all the elements of the array (X) is 1460

## mean() : -
- This is used for calculating mean of the total matrix elements.


- The numpy.mean() function is used to compute the arithmetic mean along the specified axis.


- The sum of elements, along with an axis divided by the number of elements, is known as arithmetic mean. 


- There is the following syntax of numpy.mean() function :


- <b> numpy.mean(arr, axis=None, dtype=None)
    
    
- Where, 
- i) arr : This is the source array whose elements whose mean is desired.


- ii) axis (optional) : This parameter defines the axis along which the means are computed. By default, the mean is computed of the flattened array, column-wise (axis=0) & Row-wise (axis=1).


- iii) dtype (optional) : This parameter is used to define the data-type used in computing the mean.

In [5]:
# Option - 1
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([154, 139, 154, 192, 180, 140, 154, 155, 192])

# Calculate the mean
# Print the result
np.mean(x)    # It will print the mean value of the array.

162.22222222222223

In [4]:
# Option - 2
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([154, 139, 154, 192, 180, 140, 154, 155, 192])

# Calculate the mean
# The sum of elements, along with an axis divided by the number of elements.
# Print the result
np.sum(x)/len(x)    # It will print the mean value of the array.

162.22222222222223

<b> In the above example the mean of all the elements of the array (X) is 162.22

## median() : -
- This is used for calculating / obtaining median of entire matrix elements.

- Median is nothing but sorting the given data in ascending order and select middle element.

- If the number sorted elements are odd then center or middle element becomes median.

- If the number sorted elements are even then select center or middle of two elements, add them and divided by 2 and that 
  result becomes median.


- There is the following syntax of numpy.median() function :


- <b> numpy.median(arr, axis=None, dtype=None)
    
    
- Where, 
- i) arr : This is the source array whose elements whose median is desired.


- ii) axis (optional) : This parameter defines the axis along which the median are computed. By default, the median is computed of the flattened array, column-wise (axis=0) & Row-wise (axis=1).


- iii) dtype (optional) : This parameter is used to define the data-type used in computing the median.

In [6]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([154, 139, 154, 192, 180, 140, 154, 155, 192])

# Calculate the median
# Print the result
np.median(x)    # It will print the median value of the array.

154.0

<b> In the above example the median of all the elements of the array (X) is 154.

## mode() : -
- The mode can be defined as the number that occurs most often. In other words, a number that appears most often in a set. A 
  number that has the highest frequency (repetition of a number) among the numbers, is called mode.
  
  
- If there is no repeated number in the list or observations, there is no mode for the list.


- There is the following syntax of numpy.mode() function :


- <b> numpy.mode(arr, axis=None, dtype=None)
    
    
- Where, 
- i) arr : This is the source array whose elements whose mode is desired.


- ii) axis (optional) : This parameter defines the axis along which the mode are computed. By default, the mode is computed of the flattened array, column-wise (axis=0) & Row-wise (axis=1).


- iii) dtype (optional) : This parameter is used to define the data-type used in computing the mode.

In [10]:
# importe numpy with alias name 'np'.
import numpy as np

# import the mode from scipy.stats
from scipy.stats import mode

# create an array 'x' using np.array() function.
x = np.array([154, 139, 154, 192, 180, 140, 154, 155, 192])

# Calculate the mode
# Print the result
mode(x)    # It will print the mode value of the array.

C:\Users\Administrator\AppData\Local\Temp\ipykernel_7568\2107045519.py:12: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode(x)    # It will print the mode value of the array.


ModeResult(mode=array([154]), count=array([3]))

<b> In the above example the mode of the array (X) is 154 and it occurs 3 times. In numpy module mode function is not available. So we are using scipy.stats module for importing mode function.

## min() : -
- The min() function is used to get the smallest element from the collection.


- There is the following syntax of numpy.min() function :


- <b> numpy.min(arr, axis=None)
    
    
- Where, 
- i) arr : This is the source array whose elements whose smallest element is desired.


- ii) axis (optional) : This parameter defines the axis along which the smallest element is computed. By default, the min is computed of the flattened array, column-wise (axis=0) & Row-wise (axis=1).

In [11]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([154, 139, 154, 192, 180, 140, 154, 155, 192])

# Calculate the minimum value / elements.
# Print the result
np.min(x)    # It will print the minimum value of the array.

139

<b> In the above example the minimum value of the array (X) is 139.

In [12]:
# index of the minimum element / value.
np.argmin(x)     # It will print the index of the minimum element / value.

1

## max() : -
- The max() function is used to get the largest element from the collection.


- There is the following syntax of numpy.max() function :


- <b> numpy.max(arr, axis=None)
    
    
- Where, 
- i) arr : This is the source array whose elements whose largest element is desired.


- ii) axis (optional) : This parameter defines the axis along which the largest element is computed. By default, the max is computed of the flattened array, column-wise (axis=0) & Row-wise (axis=1).

In [13]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([154, 139, 154, 192, 180, 140, 154, 155, 192])

# Calculate the maximum value / elements.
# Print the result
np.max(x)    # It will print the maximum value of the array.

192

<b> In the above example the maximum value of the array (X) is 192.

## Range : -
- Calculating the difference between largest element (max) to smallest element (min).


In [17]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([154, 139, 154, 192, 180, 140, 154, 155, 192])

# Calculate the range.
# Print the result
np.max(x)-np.min(x)

53

## Mean Absolute Deviation [abs()]: -
- The python abs() function is used to return absolute value of a number. It takes only one argument, a number whose absolute 
  value is to be returned.
  
  
- There is the following syntax of abs() function :


- <b> abs(arr)
    
    
- Where, 
- i) arr : This is the source array whose elements whose absolute value of a number is desired.


In [25]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([5, 9, 16, 17, 18])

# Calculate the deviation.
deviation = x - np.mean(x)

# Calculate the absolute deviation.
abs_deviation = abs(deviation)

# Print the absolute deviation result
abs_deviation 

array([8., 4., 3., 4., 5.])

## Variance or Mean Square Deviation (var()): -
- Calculating the arithmetic variance of the array elements along with the specified axis or multiple axes.


- By default numpy calculate the population variance only. 


- There is the following syntax of numpy.var() function :


- <b> numpy.var(arr, axis=None)
    
    
- Where, 
- i) arr : This is the source array whose elements whose variance of the array elements are desired.


- ii) axis (optional) : This parameter defines the axis along which the variance of the array elements are computed. By default, the variance of the array elements are computed of the flattened array, column-wise (axis=0) & Row-wise (axis=1).

In [28]:
#Option - 1
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([5, 9, 16, 17, 18])

# Calculate the Variance / Mean Square Deviation by using var()
np.var(x)    # It will print the population variance

26.0

In [30]:
#Option - 2
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([5, 9, 16, 17, 18])

# Calculate the deviation.
deviation = x - np.mean(x)

# Calculate the square deviation.
square_deviation = deviation * deviation

# Calculate the Variance / Mean Square Deviation
# Print the Mean Square Deviation result
np.mean(square_deviation)   # It will print the population variance


26.0

- <b> In the above examples we get population variance for the given data set. If we want to calculate the sample variance, we need to calculate as follows :

In [34]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([5, 9, 16, 17, 18])

# Calculate the sample Variance / Mean Square Deviation 
sample_variance = np.var(x) * len(x)/(len(x)-1) 

# Print the sample Variance / Mean Square Deviation  result
sample_variance   # It will print the Sample variance

32.5

## Standard Deviation (std()) : -
- This function returns the standard deviation of the array elements.


- The square root of the average square deviation (computed from the mean), is known as the standard deviation.


- By default numpy calculate the population standard deviation only.


- There is the following syntax of numpy.mode() function :


- <b> numpy.std(arr, axis=None, dtype=None)
    
    
- Where, 
- i) arr : This parameter defines the source array whose elements standard deviation is calculated.


- ii) axis (optional) : It is the axis along which the standard deviation is calculated. The standard deviation of the flattened array is computed by default, column-wise (axis=0) & Row-wise (axis=1).


- iii) dtype (optional) : This parameter defines the data type, which is used in computing the standard deviation. By default, the data type is float64 for integer type arrays, and, for float types array, it will be the same as the array type.

In [37]:
#Option - 1
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([5, 9, 16, 17, 18])

# Calculate the Variance by using var() and store it in var object.
var = np.var(x)    

# Calculate the population standard deviation without using std()
# Applying square root to the "var" object
std_deviation = np.sqrt(var)

# Print the population standard deviation result
std_deviation   # It will print the population standard deviation result

5.0990195135927845

In [38]:
#Option - 1
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([5, 9, 16, 17, 18])

# Calculate the Variance by using var() and store it in var object.
var = np.var(x)    

# Calculate the population standard deviation by using std()
# Print the population standard deviation result
np.std(x)


5.0990195135927845

- <b> In the above examples we get population standard deviation for the given data set. If we want to calculate the sample standard deviation, we need to calculate as follows :

In [42]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'x' using np.array() function.
x = np.array([5, 9, 16, 17, 18])

# Calculate the sample Variance without var() and store in "sample_variance" object
sample_variance = np.var(x) * len(x)/(len(x)-1) 

# Calculate the sample standard deviation without using std()
# Applying the square root to the "sample_variance" object
sample_std = np.sqrt(sample_variance)

# Print the sample standard deviation result
sample_std

5.70087712549569

## Percentile() : -
- percentile()function used to compute the nth percentile of the given data (array elements) along the specified axis. 


- There is the following syntax of numpy.mode() function :


- <b> numpy.percentile(arr, n, axis=None) 
    
    
- Where, 
- i) arr :input array. 
    
    
- ii) n : percentile value. 
    
    
- iii) axis : axis along which we want to calculate the percentile value. Otherwise, it will consider arr to be flattened(works on all the axis). axis = 0 means along the column and axis = 1 means working along the row. 

In [44]:
# importe numpy with alias name 'np'.
import numpy as np

# create an array 'a' using np.array() function.
a = np.array([20, 2, 7, 1, 34])

# Print the Contents of "a" array on console.
print("Contents of 'a' array : ", a)

# Calculate the 50th percentile and print the result on console.
print("50th percentile of a : ", np.percentile(a, 50))

# Calculate the 25th percentile and print the result on console.
print("25th percentile of a : ", np.percentile(a, 25))

# Calculate the 75th percentile and print the result on console.
print("75th percentile of a : ", np.percentile(a, 75))

Contents of 'a' array :  [20  2  7  1 34]
50th percentile of a :  7.0
25th percentile of a :  2.0
75th percentile of a :  20.0
